lib/WorkDoneProgress/WorkDoneProgressNotifier.php

```diff
@@ -13,46 +15,78 @@ final class WorkDoneProgressNotifier implements ProgressNotifier
      */
     private $api;
 
-    public function __construct(ClientApi $api)
-    {
-        $this->api = $api->workDoneProgress();
-    }
-
     /**
-     * {@inheritDoc}
+     * @var WorkDoneToken|null
      */
-    public function create(WorkDoneToken $token): Promise
+    private $token;
+
+    public function __construct(ClientApi $api, ?WorkDoneToken $token = null)
     {
-        return $this->api->create($token);
+        $this->api = $api->workDoneProgress();
+        $this->token = $token;
+
+        if (!$this->token) {
+            $this->create();
+        }
     }
 
     /**
      * {@inheritDoc}
      */
     public function begin(
-        WorkDoneToken $token,
         string $title,
         ?string $message = null,
         ?int $percentage = null,
         ?bool $cancellable = null
     ): void {
-        $this->api->begin($token, $title, $message, $percentage, $cancellable);
+        if (!$this->token) {
+            return;
+        }
+
+        $this->api->begin($this->token, $title, $message, $percentage, $cancellable);
     }
 
     /**
      * {@inheritDoc}
      */
     public function report(
-        WorkDoneToken $token,
         ?string $message = null,
         ?int $percentage = null,
         ?bool $cancellable = null
     ): void {
-        $this->api->report($token, $message, $percentage, $cancellable);
+        if (!$this->token) {
+            return;
+        }
+
+        $this->api->report($this->token, $message, $percentage, $cancellable);
     }
 
-    public function end(WorkDoneToken $token, ?string $message = null): void
+    public function end(?string $message = null): void
     {
-        $this->api->end($token, $message);
+        if (!$this->token) {
+            return;
+        }
+
+        $this->api->end($this->token, $message);
+        $this->token = null;
+    }
+
+    /**
+     * @throws RuntimeException if the client respond an error, contains the code & message from the response error.
+     */
+    private function create(): void
+    {
+        $token = WorkDoneToken::generate();
+        /** @var ResponseMessage $response */
+        $response = wait($this->api->create($token));
+
+        if ($response->error) {
+            throw new RuntimeException(
+                $response->error->message,
+                $response->error->code,
+            );
+        }
+
+        $this->token = $token;
     }
```

Comment:
As expected `wait` does not work.
I'm not sure how to have a synchronous behavior here, any idea ?

I could try to use `Promise::onResolve` in each methods to only send the notification after the responses (if it's a success).
But then we can't fallback to a `MessageProgressNotifier`, maybe it's OK since it's because the client didn't responded but then I would like to log it.

Reply:
I tried a solution using coroutines to keep the fallback to classic message notification if the client respond an error when initiating the progress.
I'm not convinced, the tests are still passing and it really works but it feels like the `WorkDoneProgressNoifier` do most of the heavy lifting when I wanted the factory to deal with it and each implementation to be as simple as possible so they are easy to maintain if the spec evolves or if we want to introduce some more checks (like not sending begin multiple times or stuff).

But I don't really see a way around it without loosing the fallback in case of error so the options are:
- We loose the fallback to things simple, the end use loose some informative messages through
- We find a way to wait for the response in the factory
  But then it will pause the execution which might be disturbing for the user experience